src/transformers/models/mixtral/modeling_mixtral.py

```diff
@@ -813,6 +786,80 @@ def __init__(self, *args, **kwargs):
+
+        self.act_fn = ACT2FN[config.hidden_act]
+
+    def forward(
+        self, hidden_states: torch.Tensor, selected_experts: torch.Tensor, routing_weights: torch.Tensor
+    ) -> torch.Tensor:
+        """_summary_
+
+        Args:
+            hidden_states (torch.Tensor): (batch_size * token_num, hidden_dim)
+            selected_experts (torch.Tensor): (batch_size * token_num, top_k)
+            routing_weights (torch.Tensor): (batch_size * token_num, top_k)
+
+        Returns:
+            torch.Tensor: _description_
+        """
+
+        ts, tk = hidden_states.size(0), selected_experts.size(-1)
+
+        w1 = self.w1[selected_experts]  # (batch_size * token_num, top_k, ffn_dim, hidden_dim)
+        w2 = self.w2[selected_experts]  # (batch_size * token_num, top_k, hidden_dim, ffn_dim)
+        w3 = self.w3[selected_experts]  # (batch_size * token_num, ffn_dim, hidden_dim)
+
+        x1 = torch.matmul(w1, hidden_states[:, None, :, None])
+        x3 = torch.matmul(w3, hidden_states[:, None, :, None])
+        x1 = self.act_fn(x1)
+        final_hidden_states = torch.matmul(w2, x1 * x3).reshape(ts, tk, self.hidden_dim)
+        final_hidden_states = final_hidden_states * routing_weights[:, :, None]
+        final_hidden_states = final_hidden_states.sum(dim=1)
+        return final_hidden_states
+
+
+class MixtralMoeBlock(nn.Module):
+    def __init__(self, config) -> None:
+        super().__init__()
+        self.hidden_dim = config.hidden_size
+        self.ffn_dim = config.intermediate_size
+        self.num_experts = config.num_local_experts
+        self.top_k = config.num_experts_per_tok
+
+        # gating
+        self.gate = nn.Linear(self.hidden_dim, self.num_experts, bias=False)
+        self.experts = MixtralBlockTop2MLP(config)
+        # Jitter parameters
+        self.jitter_noise = config.router_jitter_noise
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        batch_size, sequence_length, hidden_dim = hidden_states.shape
+        if self.training and self.jitter_noise > 0:
+            hidden_states *= torch.empty_like(hidden_states).uniform_(1.0 - self.jitter_noise, 1.0 + self.jitter_noise)
+        hidden_states = hidden_states.view(-1, hidden_dim)
+        # router_logits: (batch * sequence_length, n_experts)
+        router_logits = self.gate(hidden_states)
+
+        routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+        routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+        # we cast back to the input dtype
+        routing_weights = routing_weights.to(hidden_states.dtype)
+        final_hidden_states = self.experts(hidden_states, selected_experts, routing_weights)
+        final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
+        return final_hidden_states, router_logits
+
+
```

Comment:
this gathers experts for tokens, and it actually works for `torch.compile` with fullgraph and  cudagraphs support, and I think it works best when we are doing decoding phase where the batchsize is small, but it will uses more memory because we need to gather expert weights for every token if batchsize is large

Reply:
If we are supporting fast generation, then I think it's good to have this than the current version because we definitely will gain more speedups especially when decoding @ArthurZucker

Reply:
Yeah GPTFast has similar changes! I think it's super interesting but too breaking as you mention